cpp/dolfinx/fem/Expression.h

```diff
@@ -28,29 +28,45 @@ class Constant;
 /// at quadrature points in all cells. This evaluated gradient can then
 /// be used as input in to a non-FEniCS function that calculates a
 /// material constitutive model.
+
 template <typename T>
 class Expression
 {
 public:
   /// Create an Expression
   ///
+  /// Users should prefer the create_expression factory functions
+  ///
   /// @param[in] coefficients Coefficients in the Expression
   /// @param[in] constants Constants in the Expression
   /// @param[in] mesh
-  /// @param[in] X points on reference cell, shape=(num_points, tdim)
+  /// @param[in] X points on reference cell, number of points rows and
+  /// tdim cols
   /// @param[in] fn function for tabulating expression
-  /// @param[in] value_size size of expression evaluated at single point
+  /// @param[in] value_shape shape of expression evaluated at single point
+  /// @param[in] argument_function_space Function space for Argument
   Expression(
       const std::vector<std::shared_ptr<const Function<T>>>& coefficients,
       const std::vector<std::shared_ptr<const Constant<T>>>& constants,
-      const std::shared_ptr<const mesh::Mesh>& mesh,
       const xt::xtensor<double, 2>& X,
-      const std::function<void(T*, const T*, const T*, const double*)> fn,
-      std::size_t value_size)
-      : _coefficients(coefficients), _constants(constants), _mesh(mesh), _x(X),
-        _fn(fn), _value_size(value_size)
+      const std::function<void(T*, const T*, const T*, const double*,
+                               const int*, const uint8_t*)>
+          fn,
+      const std::vector<int>& value_shape,
+      const std::shared_ptr<const mesh::Mesh> mesh = nullptr,
+      const std::shared_ptr<const FunctionSpace> argument_function_space
+      = nullptr)
+      : _coefficients(coefficients), _constants(constants), _mesh(mesh),
+        _x_ref(X), _fn(fn), _value_shape(value_shape),
+        _argument_function_space(argument_function_space)
   {
-    // Do nothing
+    // Extract mesh from argument's function space
+    if (!_mesh and argument_function_space)
+      _mesh = argument_function_space->mesh();
+    else if (argument_function_space and _mesh != argument_function_space->mesh())
```

Comment:
The logic here around `_mesh` is a bit confusing. I would suggest making the `else if` a separate check before extracting. The first `if` changes `_mesh` and the second check involves `_mesh` but doesn't see the possible change. Correct but potentially confusing.